sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala

```diff
@@ -1402,28 +1402,27 @@ private[rapids] class GpuRangeIterator(
           }
         }
       }
-      assert(iter.hasNext)
-      closeOnExcept(iter.next()) { batch =>
-        // This "iter" returned from the "withRetry" block above has only one batch,
-        // because the split function "reduceRowsNumberByHalf" returns a Seq with a single
-        // element inside.
-        // By doing this, we can pull out this single batch directly without maintaining
-        // this extra `iter` for the next loop.
-        assert(iter.isEmpty)
-        val endInclusive = start + ((batch.numRows() - 1) * step)
-        currentPosition = endInclusive + step
-        if (currentPosition < endInclusive ^ step < 0) {
-          // we have Long.MaxValue + Long.MaxValue < Long.MaxValue
-          // and Long.MinValue + Long.MinValue > Long.MinValue, so iff the step causes a
-          // step back, we are pretty sure that we have an overflow.
-          done = true
-        }
-        if (batch.numRows() < rowsExpected) {
-          logDebug(s"Retried with ${batch.numRows()} rows when expected $rowsExpected rows")
+      if (iter.hasNext) {
+        closeOnExcept(iter.next()) { batch =>
+          // This "iter" returned from the "withRetry" block above has only one batch,
+          // because the split function "reduceRowsNumberByHalf" returns a Seq with a single
+          // element inside.
+          // By doing this, we can pull out this single batch directly without maintaining
+          // this extra `iter` for the next loop.
+          val endInclusive = start + ((batch.numRows() - 1) * step)
+          currentPosition = endInclusive + step
+          if (currentPosition < endInclusive ^ step < 0) {
+            // we have Long.MaxValue + Long.MaxValue < Long.MaxValue
+            // and Long.MinValue + Long.MinValue > Long.MinValue, so iff the step causes a
+            // step back, we are pretty sure that we have an overflow.
+            done = true
+          }
+          if (batch.numRows() < rowsExpected) {
+            logDebug(s"Retried with ${batch.numRows()} rows when expected $rowsExpected rows")
+          }
+          batch
```

Comment:
there is no else here, so this should fail to compile. I would: 

1. revert this piece
2. remove the old `assert(iter.hasNext)` in the old line 1405, as that seems redundant.
3. turn the old `assert(iter.isEmpty)` into a `requires(iter.isEmpty, "a single batch was expected, but retry iterator was not exhausted")`.